### ISSUE TITLE
Source CIDR/mask, ranges, support in multicast rules

### ISSUE THREAD

**Issue by gedl** [troglobit/smcroute#81]
I tried to look around open and old issues, but couldn't find any reference to this.

I would like to specify a rule to include multiple hosts. Let's take the following as an example:

`smcroute -a i1 172.20.2.2 239.255.255.250 i2`

I need this rule to be repeated for a number of source IPs. Let's say I can specified this range of IPs as `172.20.2.2/24`.

Is there any way of specifying a CIDDR or a range or a mask when adding a rule to smcroute? The companion iptables (mangle table) to increment the TTL is easy, but couldn't find one for smcroute.

Basically I need to `send multicast from IPs 172.20.2.0.0 - 172.20.2.0.255 coming from it1 to 239.255.255.250 through i2`

Am I missing something ?

**Comment by troglobit** [comment]
You're not missing anything.

The feature you're looking for is not available yet. I recently added support for destination group ranges. Adding support for source/sender ranges should be fairly simple.

It is worth noting, however, that there is limited kernel support for these features so SMCRoute has to basically create single one-to-one mappings to feed the kernel.

**Comment by gedl** [comment]
Understood, that's what I'm doing right now: my setup script loops through each address and issues a `smcroute -a` command for each IP in the specified range.

**Comment by kylex2** [comment]
First, thanks for making SMCRoute!
I wanted to try these kind of routes so I had a [go](https://github.com/kylex2/smcroute/commit/4bc1496066028febf0fe43bf64103fdd8ead40e9) at the source code. I am sure I have missed stuff so any comments are welcome.

**Comment by troglobit** [comment]
@kylex2 There are lot's of unrelated changes in that patch (whitespace, changes to log messages) that would not be accepted in a pull request.  However, in general your patch looks set to cover the basics of what op wanted.

I cannot comment more without downloading and testing it, but I'd like to request you do a formal pull request for that.  In which case I'd very much like you to split up the changes into logically separate commits and not include unrelated changes.  We can also do a proper code audit in the PR, but as a heads-up, the first `strchr()` can segfault since `source` may be NULL.

**Comment by troglobit** [comment]
This is now fixed and on master for the v2.5 release thanks to the awesome work of @kylex2 :-)

**Comment by troglobit** [comment]
Also, we'd greatly appreciate all feedback and testing of this you guys can contribute with. Thanks!
